File: typescript/tests/commerce_xero_board.test.ts
// commerce_xero_board.test.ts
// Verifies the CircleAI.Commerce.Integration.Xero port: token storage/expiry,
// tenant de-dup, and webhook recording (recent-first).

import { describe, it } from "node:test";
import assert from "node:assert/strict";
import {
  InMemoryXeroBoard,
  CommerceIntegrationXeroDomainContext,
  xeroTokens,
  xeroTenant,
  xeroWebhookEvent,
} from "../src/commerce/integration/xero/index";

describe("InMemoryXeroBoard — tokens", () => {
  it("stores/retrieves tokens and reports expiry vs now", () => {
    const b = new InMemoryXeroBoard();
    const exp = new Date("2026-07-10T12:00:00Z");
    b.storeTokens("u1", xeroTokens("acc", "ref", exp, "id"));
    assert.equal(b.getTokens("u1")?.accessToken, "acc");
    assert.equal(b.tokensExpired("u1", new Date("2026-07-10T11:59:59Z")), false);
    assert.equal(b.tokensExpired("u1", new Date("2026-07-10T12:00:00Z")), true); // now >= expiry
    assert.equal(b.tokensExpired("u1", new Date("2026-07-10T12:00:01Z")), true);
  });

  it("unknown user is treated as expired and has no tokens", () => {
    const b = new InMemoryXeroBoard();
    assert.equal(b.getTokens("ghost"), undefined);
    assert.equal(b.tokensExpired("ghost", new Date()), true);
  });
});

describe("InMemoryXeroBoard — tenants", () => {
  it("adds tenants per user and de-duplicates by TenantId", () => {
    const b = new InMemoryXeroBoard();
    b.addTenant("u1", xeroTenant("t1", "Org1", "ORGANISATION"));
    b.addTenant("u1", xeroTenant("t1", "Org1-dup", "ORGANISATION")); // same id → ignored
    b.addTenant("u1", xeroTenant("t2", "Org2", "ORGANISATION"));
    assert.deepEqual(
      b.tenantsFor("u1").map((t) => t.tenantId),
      ["t1", "t2"],
    );
    assert.deepEqual(b.tenantsFor("other"), []);
  });
});

describe("InMemoryXeroBoard — webhooks", () => {
  it("records events and returns them most-recent-first (by AtUtc), limited", () => {
    const b = new InMemoryXeroBoard();
    b.recordWebhook(xeroWebhookEvent("t1", "Invoice", "r1", new Date("2026-01-01T00:00:00Z")));
    b.recordWebhook(xeroWebhookEvent("t1", "Invoice", "r3", new Date("2026-03-01T00:00:00Z")));
    b.recordWebhook(xeroWebhookEvent("t1", "Invoice", "r2", new Date("2026-02-01T00:00:00Z")));
    assert.deepEqual(
      b.recentEvents(2).map((e) => e.resourceId),
      ["r3", "r2"],
    );
    assert.equal(b.recentEvents().length, 3);
  });

  it("domain context exposes prompt + compliance + tools", () => {
    assert.ok(CommerceIntegrationXeroDomainContext.systemPromptSnippet.includes("[DOMAIN: Commerce.Integration.Xero]"));
    assert.deepEqual(CommerceIntegrationXeroDomainContext.complianceFlags, ["SARS", "IFRS", "Xero_Data_Standards", "POPIA"]);
    assert.deepEqual(CommerceIntegrationXeroDomainContext.suggestedTools, ["xero_api", "spreadsheet", "document_editor"]);
  });
});

File: go/tests/commerce_xero_board_test.go
// commerce_xero_board_test.go
//
// Verifies the CircleAI.Commerce.Integration.Xero port (commerce_xero_board.go):
// token store/get, expiry (missing => expired), tenant de-dup + insertion order,
// and reverse-chronological recent-event recall with cap.

package circleai_test

import (
	"testing"
	"time"

	circleai "github.com/bhengubv/CircleAI/go"
)

func TestXero_TokensStoreGetExpiry(t *testing.T) {
	b := circleai.NewInMemoryXeroBoard()
	exp := time.Date(2026, 7, 11, 12, 0, 0, 0, time.UTC)
	b.StoreTokens("u1", circleai.XeroTokens{AccessToken: "at", RefreshToken: "rt", ExpiresAtUtc: exp, IdToken: "id"})
	if tok, ok := b.GetTokens("u1"); !ok || tok.AccessToken != "at" {
		t.Fatalf("get tokens = %+v ok=%v", tok, ok)
	}
	if _, ok := b.GetTokens("nobody"); ok {
		t.Fatalf("missing tokens found")
	}
	// Missing user => expired.
	if !b.TokensExpired("nobody", exp) {
		t.Fatalf("missing user should be expired")
	}
	// Before expiry => not expired.
	if b.TokensExpired("u1", exp.Add(-time.Hour)) {
		t.Fatalf("before expiry should not be expired")
	}
	// At/after expiry => expired (now >= ExpiresAtUtc).
	if !b.TokensExpired("u1", exp) {
		t.Fatalf("at expiry should be expired")
	}
	if !b.TokensExpired("u1", exp.Add(time.Hour)) {
		t.Fatalf("after expiry should be expired")
	}
}

func TestXero_TenantDedupAndOrder(t *testing.T) {
	b := circleai.NewInMemoryXeroBoard()
	b.AddTenant("u1", circleai.XeroTenant{TenantId: "t1", TenantName: "Org One", TenantType: "ORGANISATION"})
	b.AddTenant("u1", circleai.XeroTenant{TenantId: "t2", TenantName: "Org Two", TenantType: "ORGANISATION"})
	b.AddTenant("u1", circleai.XeroTenant{TenantId: "t1", TenantName: "Dup", TenantType: "ORGANISATION"}) // dup ignored
	tenants := b.TenantsFor("u1")
	if len(tenants) != 2 || tenants[0].TenantId != "t1" || tenants[1].TenantId != "t2" {
		t.Fatalf("tenant dedup/order failed: %+v", tenants)
	}
	if tenants[0].TenantName != "Org One" {
		t.Fatalf("duplicate should not overwrite: %q", tenants[0].TenantName)
	}
	if len(b.TenantsFor("nobody")) != 0 {
		t.Fatalf("unknown user tenants should be empty")
	}
}

func TestXero_RecentEvents(t *testing.T) {
	b := circleai.NewInMemoryXeroBoard()
	base := time.Date(2026, 7, 1, 0, 0, 0, 0, time.UTC)
	b.RecordWebhook(circleai.XeroWebhookEvent{TenantId: "t1", ResourceType: "Invoice", ResourceId: "r1", AtUtc: base})
	b.RecordWebhook(circleai.XeroWebhookEvent{TenantId: "t1", ResourceType: "Invoice", ResourceId: "r3", AtUtc: base.Add(2 * time.Hour)})
	b.RecordWebhook(circleai.XeroWebhookEvent{TenantId: "t1", ResourceType: "Invoice", ResourceId: "r2", AtUtc: base.Add(time.Hour)})

	recent := b.RecentEvents(20)
	if len(recent) != 3 || recent[0].ResourceId != "r3" || recent[1].ResourceId != "r2" || recent[2].ResourceId != "r1" {
		t.Fatalf("recent events desc failed: %+v", recent)
	}
	if capped := b.RecentEvents(1); len(capped) != 1 || capped[0].ResourceId != "r3" {
		t.Fatalf("capped recent events failed: %+v", capped)
	}
}
